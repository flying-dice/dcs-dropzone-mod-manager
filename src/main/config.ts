import { basename, join } from 'node:path'
import { app } from 'electron'
import { Logger } from '@nestjs/common'
import { DEFAULT_RCLONE_URL, DEFAULT_REGISTRY_URL } from './constants'
import { findInstalledDcsWriteDir } from './functions/findInstalledDcsWriteDir'
import { z } from 'zod'

const mainConfigSchema = z.object({
  tempDir: z.string(),
  writeDir: z.string(),
  defaultRegistryUrl: z.string(),
  appDataName: z.string(),
  defaultDcsWriteDir: z.string().optional(),
  toolsDir: z.string(),
  logfile: z.string(),
  logfileDisplayName: z.string().optional(),
  mongo: z.object({
    port: z.number(),
    dbPath: z.string()
  }),
  rcloneInstance: z.object({
    baseURL: z.string()
  })
})

export type MainConfig = z.infer<typeof mainConfigSchema>

export async function configuration(): Promise<MainConfig> {
  const appDataName = basename(app.getPath('userData'))

  const logfile = join(app.getPath('logs'), 'main.log')
  const logfileDisplayName = join(app.getPath('logs'), 'main.log').replace(
    app.getPath('home'),
    '%USERPROFILE%'
  )

  const tempDir = join(app.getPath('temp'), appDataName)
  const writeDir = join(process.env.LOCALAPPDATA || app.getPath('userData'), 'dcs-dropzone')
  const toolsDir = writeDir

  Logger.log(tempDir, 'tempDir')
  Logger.log(writeDir, 'writeDir')
  Logger.log(toolsDir, 'toolsDir')
  Logger.log(appDataName, 'appDataName')
  Logger.log(logfile, 'logfile')

  const defaultDcsWriteDir = findInstalledDcsWriteDir()

  const config: MainConfig = {
    defaultRegistryUrl: DEFAULT_REGISTRY_URL,
    tempDir,
    writeDir,
    toolsDir,
    appDataName,
    logfile,
    logfileDisplayName,
    defaultDcsWriteDir,
    rcloneInstance: {
      baseURL: DEFAULT_RCLONE_URL
    },
    mongo: {
      port: 57449,
      dbPath: join(app.getPath('userData'), '__data')
    }
  }

  return mainConfigSchema.parse(config)
}
