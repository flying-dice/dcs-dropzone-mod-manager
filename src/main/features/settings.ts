import { join } from 'path'
import { dialog } from 'electron'
import { pathExistsSync } from 'fs-extra'
import { trpc } from '../trpc'


export const settingsService = {
  async getDefaultSaveGameDir(): Promise<{ path: string; valid: boolean }> {
    if (!process.env.USERPROFILE) {
      return {
        path: '',
        valid: false
      }
    }

    const defaultPath = join(process.env.USERPROFILE, 'Saved Games', 'DCS')
    const defaultOBPath = join(process.env.USERPROFILE, 'Saved Games', 'DCS.openbeta')

    const paths = [defaultPath, defaultOBPath]

    const validPath = paths.find((it) => pathExistsSync(it))

    if (validPath) {
      return {
        path: validPath,
        valid: true
      }
    }
    return {
      path: defaultPath,
      valid: false
    }
  },
  async getSaveGameDir(): Promise<string> {
    const defaultPath = await settingsService.getDefaultSaveGameDir()
    return await dialog
      .showOpenDialog({
        defaultPath: defaultPath?.path,
        properties: ['openDirectory']
      })
      .then((it) => it.filePaths[0])
  },
  async getDefaultWriteDir(): Promise<{ path: string; valid: boolean }> {
    if (!process.env.USERPROFILE) {
      return {
        path: '',
        valid: false
      }
    }
  
    const defaultPath = join(process.env.USERPROFILE, 'Saved Games' ,'Dropzone', 'Mods')
  
    return {
      path: defaultPath,
      valid: pathExistsSync(defaultPath)
    }

  },
  async getWriteDir(): Promise<string> {
    const defaultPath = await settingsService.getDefaultWriteDir()
    return await dialog
      .showOpenDialog({
        defaultPath: defaultPath?.path,
        properties: ['openDirectory']
      })
      .then((it) => it.filePaths[0])
  },
}



export const settingsRouter = trpc.router({
  getWriteDir: trpc.procedure.query(settingsService.getWriteDir),
  getDefaultWriteDir: trpc.procedure.query(settingsService.getDefaultWriteDir),
  getDefaultSaveGameDir: trpc.procedure.query(settingsService.getDefaultSaveGameDir),
})
