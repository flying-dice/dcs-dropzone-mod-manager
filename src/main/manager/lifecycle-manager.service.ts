import { existsSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { Inject, Injectable, Logger } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import { ensureDirSync, rm, symlink } from 'fs-extra'
import type { Repository } from 'typeorm'
import { ReleaseAssetEntity } from '../entities/release-asset.entity'
import { ReleaseEntity } from '../entities/release.entity'
import { SubscriptionEntity } from '../entities/subscription.entity'
import type { FsService } from '../services/fs.service'
import type { WriteDirectoryService } from '../services/write-directory.service'
import { HashPath } from '../utils/hash-path'
import type { SettingsManager } from './settings.manager'

/**
 * Manages the toggling of a mod between enabled and disabled states
 *
 * Also exposes methods to query the state of assets for a mod
 */
@Injectable()
export class LifecycleManager {
  private readonly logger = new Logger(LifecycleManager.name)

  @InjectRepository(SubscriptionEntity)
  private readonly subscriptionRepository: Repository<SubscriptionEntity>

  @InjectRepository(ReleaseEntity)
  private readonly releaseRepository: Repository<ReleaseEntity>

  @InjectRepository(ReleaseAssetEntity)
  private readonly releaseAssetRepository: Repository<ReleaseAssetEntity>

  @Inject()
  private readonly writeDirectoryService: WriteDirectoryService

  @Inject()
  private readonly settingsManager: SettingsManager

  @Inject()
  private readonly fsService: FsService

  /**
   * Toggles the mod between enabled and disabled states
   * If the mod is enabled, it will be disabled
   * If the mod is disabled, it will be enabled
   */
  async toggleMod(modId: string): Promise<void> {
    const { release } = await this.getSubscriptionWithReleaseOrThrow(modId)
    this.logger.debug(`Toggling mod: ${modId} which is currently ${release.enabled}`)
    if (release.enabled) {
      this.logger.debug(`Disabling mod: ${modId}`)
      await this.disableMod(modId)
    } else {
      this.logger.debug(`Enabling mod: ${modId}`)
      await this.enableMod(modId)
    }
  }

  /**
   * Enables the mod
   * If the mod is already enabled, it will remain enabled
   */
  async enableMod(modId: string): Promise<void> {
    const { release, subscription } = await this.getSubscriptionWithReleaseOrThrow(modId)
    const releaseAssets = await this.releaseAssetRepository.findBy({ release })
    this.logger.debug(`Enabling mod: ${modId} with ${releaseAssets.length} release assets`)
    for (const releaseAsset of releaseAssets) {
      this.logger.debug(`Enabling release asset: ${releaseAsset.id}`)

      let srcPath = join(
        await this.writeDirectoryService.getWriteDirectoryForRelease(subscription.id, release.id),
        releaseAsset.source
      )

      // If the source is a hash path, we need to extract the base path and make sure the symlink is for the exploded folder including internal route
      if (HashPath.isHashPath(srcPath)) {
        this.logger.debug(`Source is a hash path: ${srcPath}`)
        const hashPath = new HashPath(srcPath)
        srcPath = join(hashPath.basePathWithoutExt, hashPath.hashPath)
      }

      const targetPath = join(await this.settingsManager.getGameDir(), releaseAsset.target)
      this.logger.debug(
        `Creating Symlink for release asset: ${releaseAsset.id} from ${srcPath} to ${targetPath}`
      )
      ensureDirSync(dirname(targetPath))
      await symlink(join(srcPath), targetPath)
      if (existsSync(targetPath)) {
        releaseAsset.symlinkPath = targetPath
        await this.releaseAssetRepository.save(releaseAsset)
      } else {
        throw new Error(`Symlink not created: ${targetPath}`)
      }
    }

    release.enabled = true
    await this.releaseRepository.save(release)
  }

  /**
   * Disables the mod
   * If the mod is already disabled, it will remain disabled
   */
  async disableMod(modId: string): Promise<void> {
    const { release } = await this.getSubscriptionWithReleaseOrThrow(modId)
    const releaseAssets = await this.releaseAssetRepository.findBy({ release })

    this.logger.debug(`Disabling mod: ${modId} with ${releaseAssets.length} release assets`)
    for (const releaseAsset of releaseAssets) {
      if (releaseAsset.symlinkPath !== null) {
        this.logger.debug(`Deleting Symlink for release asset: ${releaseAsset.id}`)
        await rm(releaseAsset.symlinkPath, { force: true, recursive: true })
        releaseAsset.symlinkPath = null
        await this.releaseAssetRepository.save(releaseAsset)
      }
    }
    release.enabled = false
    await this.releaseRepository.save(release)
  }

  /**
   * Gets the subscription and release for the mod or throws an error if not found
   *
   * @param modId
   * @private
   */
  private async getSubscriptionWithReleaseOrThrow(modId: string) {
    const subscription = await this.subscriptionRepository.findOneByOrFail({
      modId
    })
    const release = await this.releaseRepository.findOneByOrFail({
      subscription
    })

    return { subscription, release }
  }

  async getModAssets(modId: string) {
    const { release } = await this.getSubscriptionWithReleaseOrThrow(modId)
    const releaseAssets = await this.releaseAssetRepository.findBy({ release })
    return releaseAssets.map((it) => ({
      id: it.id,
      source: it.source,
      symlinkPath: it.symlinkPath
    }))
  }

  async openAssetInExplorer(assetId: number) {
    const asset = await this.releaseAssetRepository.findOneByOrFail({
      id: assetId
    })
    if (!asset.symlinkPath) throw new Error(`Symlink path not present, is the mod enabled?`)
    this.logger.debug(`Opening asset in explorer: ${asset.id}`)

    return this.fsService.openFolder(dirname(asset.symlinkPath))
  }
}
