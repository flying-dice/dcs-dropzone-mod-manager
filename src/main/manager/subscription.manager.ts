import { Inject, Injectable, Logger, OnApplicationBootstrap } from '@nestjs/common'
import { ProgressLabel } from '../../lib/types'
import { FsService } from '../services/fs.service'
import { RegistryService } from '../services/registry.service'
import { WriteDirectoryService } from '../services/write-directory.service'
import { LifecycleManager } from './lifecycle-manager.service'
import { trackEvent } from '@aptabase/electron/main'
import { Subscription } from '../schemas/subscription.schema'
import { Release } from '../schemas/release.schema'
import { SubscriptionService } from '../services/subscription.service'
import { ReleaseService } from '../services/release.service'
import { Log } from '../utils/log'
import { getReleaseAsset } from '../utils/get-release-asset'
import { randomUUID } from 'node:crypto'
import { join } from 'node:path'
import { existsSync, readdirSync, ensureDirSync, rmdir, pathExistsSync } from 'fs-extra'
import { AssetTaskStatus } from '../schemas/release-asset-task.schema'

export type SubscriptionReleaseState = {
  enabled: boolean
  version: string
  progressLabel: ProgressLabel
  progress: number
  currentTaskLabel?: string
  exePath?: string
  isLatest: boolean
  latest?: string
  isReady: boolean
}

export type SubscriptionWithState = {
  subscription: Subscription
  state: SubscriptionReleaseState
}

@Injectable()
export class SubscriptionManager implements OnApplicationBootstrap {
  private readonly logger = new Logger(SubscriptionManager.name)

  @Inject(SubscriptionService)
  private readonly subscriptionService: SubscriptionService

  @Inject(ReleaseService)
  private readonly releaseService: ReleaseService

  @Inject()
  private readonly registryService: RegistryService

  @Inject()
  private readonly writeDirectoryService: WriteDirectoryService

  @Inject()
  private readonly fsService: FsService

  @Inject()
  private readonly toggleManager: LifecycleManager

  async onApplicationBootstrap(): Promise<any> {
    this.logger.log('Checking for orphaned subscriptions and releases')
    await this.removeOrphanedSubscriptionsAndReleases()
  }

  @Log()
  async getAllSubscriptions(): Promise<SubscriptionWithState[]> {
    const subscriptionsWithState: SubscriptionWithState[] = []

    for (const subscription of await this.subscriptionService.findAll()) {
      const state = await this.getSubscriptionReleaseState(subscription)
      if (state) {
        subscriptionsWithState.push({ subscription, state })
      }
    }

    return subscriptionsWithState
  }

  /**
   * Gets the subscription and release for the mod or throws an error if not found
   * This is primarily used on the MyContent page to render the current subscriptions and their status
   */
  @Log()
  async getSubscriptionReleaseState(
    subscription: Subscription
  ): Promise<SubscriptionReleaseState | undefined> {
    const installed = await this.releaseService.findBySubscriptionIdOrThrow(subscription.id)
    const latest = await this.registryService.getLatestVersion(subscription.modId)

    const assetTasks = await this.releaseService.findAssetTasksByRelease(installed.id)

    const taskStatuses = assetTasks.map((it) => it.status)
    const taskProgress = assetTasks.map((it) => it.progress)

    let progressLabel: ProgressLabel = 'Pending'

    if (taskStatuses.every((it) => it === AssetTaskStatus.COMPLETED)) {
      progressLabel = 'Completed'
    }

    if (taskStatuses.some((it) => it === AssetTaskStatus.FAILED)) {
      progressLabel = 'Failed'
    }

    if (taskStatuses.some((it) => it === AssetTaskStatus.IN_PROGRESS)) {
      progressLabel = 'In Progress'
    }

    const progress = taskProgress.reduce((acc, cur) => acc + cur, 0) / taskProgress.length

    return {
      enabled: installed.enabled,
      version: installed.version,
      progressLabel,
      progress,
      currentTaskLabel: assetTasks.find((it) => it.status === AssetTaskStatus.IN_PROGRESS)?.label,
      isReady: taskStatuses.every((it) => it === AssetTaskStatus.COMPLETED),
      exePath: installed.exePath,
      isLatest: installed.version === latest?.version || true,
      latest: latest?.version
    }
  }

  @Log()
  async subscribe(modId: string): Promise<void> {
    this.logger.log(`Subscribing to mod ${modId}`)

    this.logger.debug(`Getting registry index for mod ${modId}`)
    const mod = await this.registryService.getRegistryEntryIndex(modId)

    this.logger.debug(`Getting latest release for mod ${modId}`)
    const latestRelease = await this.registryService.getLatestVersion(modId)

    if (!latestRelease) {
      throw new Error(`No releases found for mod ${modId}`)
    }

    this.logger.debug(`Building Subscription and Release data for ${modId}`)
    const subscription: Subscription = {
      id: randomUUID(),
      modId: mod.id,
      modName: mod.name!,
      created: Date.now()
    }
    const release: Release = {
      id: randomUUID(),
      subscriptionId: subscription.id,
      version: latestRelease.version!,
      enabled: false,
      exePath: latestRelease.exePath
    }

    this.logger.debug(`Determining write directory for mod ${modId}`)
    const releaseWriteDir = await this.writeDirectoryService.getWriteDirectoryForRelease(
      subscription,
      release
    )
    this.logger.verbose(`Release write directory: ${releaseWriteDir}`)

    const assets = latestRelease.assets.map((asset) =>
      getReleaseAsset(release, asset, releaseWriteDir)
    )

    this.logger.debug(`Creating write directories`)
    ensureDirSync(releaseWriteDir)

    this.logger.debug(`Saving subscription for mod ${modId}`)
    await this.subscriptionService.save(subscription)

    this.logger.debug(`Saving release ${latestRelease.version} for mod ${modId}`)
    await this.releaseService.save(release)

    this.logger.debug(`Saving release assets for mod ${modId}`)
    for (const asset of assets) {
      await this.releaseService.saveAsset(asset)
      for (const task of asset.tasks) {
        await this.releaseService.saveAssetTask(task)
      }
    }

    this.logger.log(`Subscribed to mod ${modId}`)

    await trackEvent('mod_subscribed', { mod_id: modId, mod_version: latestRelease.version! })
  }

  @Log()
  async unsubscribe(modId: string): Promise<void> {
    this.logger.debug(`Finding subscription for mod ${modId}`)
    const subscription = await this.subscriptionService.findByModId(modId)

    if (!subscription) {
      this.logger.warn(`Subscription for mod ${modId} not found`)
      return
    }

    this.logger.debug(`Disabling mod ${modId}`)
    await this.toggleManager.disableMod(modId)

    const modDir = await this.writeDirectoryService.getWriteDirectoryForSubscription(subscription)

    if (existsSync(modDir)) {
      this.logger.debug(`Deleting write directory for mod ${modId}`)
      await rmdir(modDir, { recursive: true })
    }

    this.logger.debug(`Deleting subscription for mod ${modId}`)
    await this.releaseService.deleteBySubscriptionId(subscription.id)
    await this.subscriptionService.deleteById(subscription.id)
  }

  @Log()
  async openInExplorer(modId: string): Promise<void> {
    const subscription = await this.subscriptionService.findByModId(modId)
    if (subscription) {
      const modDir = await this.writeDirectoryService.getWriteDirectoryForSubscription(subscription)
      await this.fsService.openFolder(modDir)
    }
  }

  @Log()
  async update(modId: string) {
    this.logger.log(`Updating mod ${modId}`)

    this.logger.debug(`Unsubscribing from mod ${modId}`)
    await this.unsubscribe(modId)

    this.logger.debug(`Subscribing to mod ${modId}`)
    await this.subscribe(modId)
  }

  @Log()
  async removeOrphanedSubscriptionsAndReleases() {
    const writeDirectory = await this.writeDirectoryService.getWriteDirectory()

    if (!pathExistsSync(writeDirectory)) return

    const foldersInWriteDirectory = readdirSync(writeDirectory, { withFileTypes: true }).filter(
      (it) => it.isDirectory()
    )

    for (const directory of foldersInWriteDirectory) {
      const subscription = await this.subscriptionService.findById(directory.name)
      if (!subscription) {
        this.logger.warn(`Deleting orphaned directory ${directory.name}`)
        await rmdir(join(directory.parentPath, directory.name), { recursive: true })
        continue
      }

      await this.removeOrphanedReleases(subscription)
    }
  }

  @Log()
  async removeOrphanedReleases(subscription: Subscription) {
    const subscriptionDirectory =
      await this.writeDirectoryService.getWriteDirectoryForSubscription(subscription)

    const foldersInSubscriptionDirectory = readdirSync(subscriptionDirectory, {
      withFileTypes: true
    }).filter((it) => it.isDirectory())

    for (const directory of foldersInSubscriptionDirectory) {
      const release = await this.releaseService.findById(directory.name)
      if (!release) {
        this.logger.warn(`Deleting orphaned directory ${directory.name}`)
        await rmdir(join(directory.parentPath, directory.name), { recursive: true })
      }
    }
  }
}
