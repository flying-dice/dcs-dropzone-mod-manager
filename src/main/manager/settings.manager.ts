import { Inject, Injectable } from '@nestjs/common'
import { SettingsService } from '../services/settings.service'
import { DEFAULT_REGISTRY_URL } from '../constants'
import { ConfigService } from '@nestjs/config'
import { join } from 'node:path'
import { MainConfig } from '../config'

/**
 * Settings manager for handling settings related to the application at the implementation level
 * For general access to the settings repository, see the ConfigService
 *
 * This class is mostly used to provide default values for settings that are not set,
 * i.e. the registry URL, installation folders and write directories
 */
@Injectable()
export class SettingsManager {
  @Inject(SettingsService)
  private settingsService: SettingsService

  @Inject(ConfigService)
  private configService: ConfigService<MainConfig>

  /**
   * Get the registry URL from the settings repository, or the default value if not set
   */
  async getRegistryUrl(): Promise<string> {
    const url = (await this.settingsService.getSettingValue('registryUrl'))?.value
    return url || this.getDefaultRegistryUrl()
  }

  async getDefaultRegistryUrl(): Promise<string> {
    return DEFAULT_REGISTRY_URL
  }

  /**
   * Get the write directory from the settings repository, or the default value if not set
   *
   * Defaults to the dcs-dropzone directory in the user's documents folder @see FsService.getDefaultWriteDir
   */
  async getWriteDir(): Promise<string> {
    const configuredDir = (await this.settingsService.getSettingValue('writeDir'))?.value
    return configuredDir || this.getDefaultWriteDir()
  }

  async getDefaultWriteDir(): Promise<string> {
    return join(this.configService.getOrThrow('writeDir'), 'mods')
  }

  /**
   * Get the installation directory from the settings repository, or the default value if not set
   * Defaults to the DCS installation directory
   *
   * @throws Error if no game directory is set and the default is not found
   */
  async getGameDir(): Promise<string> {
    const configuredDir = (await this.settingsService.getSettingValue('gameDir'))?.value
    const value = configuredDir || (await this.getDefaultGameDir())

    if (!value) {
      throw new Error(
        'No game directory configured and default not found, ensure DCS is installed and has been run at least once'
      )
    }

    return value
  }

  async getDefaultGameDir(): Promise<string | undefined> {
    return this.configService.get('defaultDcsWriteDir')
  }
}
