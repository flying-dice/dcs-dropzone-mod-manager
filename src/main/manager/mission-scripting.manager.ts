import { Inject, Injectable, Logger } from '@nestjs/common'
import { OnEvent } from '@nestjs/event-emitter'
import { ModEnabledEvent } from '../events/mod-enabled.event'
import { ModDisabledEvent } from '../events/mod-disabled.event'
import { ReleaseService } from '../services/release.service'
import { outputFile } from 'fs-extra'
import { SettingsManager } from './settings.manager'
import path from 'node:path'
import { generateDropzoneMissionScriptingScript } from '../functions/generateDropzoneMissionScriptingScript'

@Injectable()
export class MissionScriptingManager {
  static readonly FILENAME = 'dcs-dropzone_MissionScripting.lua'

  private readonly logger = new Logger(MissionScriptingManager.name)

  @Inject(ReleaseService)
  private readonly releaseService: ReleaseService

  @Inject()
  private readonly settingsManager: SettingsManager

  @OnEvent(ModEnabledEvent.name)
  async onModEnabled(event: ModEnabledEvent) {
    this.logger.log(`Handling mod enabled event: ${event.modId} ${event.version}`)
    await this.rebuildUninstallBat()
  }

  @OnEvent(ModDisabledEvent.name)
  async onModDisabled(event: ModDisabledEvent) {
    this.logger.log(`Handling mod disabled event: ${event.modId} ${event.version}`)
    await this.rebuildUninstallBat()
  }

  async rebuildUninstallBat() {
    const releaseAssets = await this.releaseService.findAssetsWithSymlinks()

    const scripts: { id: string; path: string }[] = []

    for (const asset of releaseAssets) {
      if (asset.symlinkPath && asset.runOnStart) {
        scripts.push({
          id: asset.id,
          path: asset.symlinkPath
        })
      }
    }

    const content = await generateDropzoneMissionScriptingScript(scripts)
    await outputFile(
      path.join(
        await this.settingsManager.getGameDir(),
        'Scripts',
        MissionScriptingManager.FILENAME
      ),
      content
    )
  }
}
