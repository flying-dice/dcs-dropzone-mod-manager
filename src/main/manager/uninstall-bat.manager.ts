import { Inject, Injectable, Logger } from '@nestjs/common'
import { OnEvent } from '@nestjs/event-emitter'
import { ModEnabledEvent } from '../events/mod-enabled.event'
import { ModDisabledEvent } from '../events/mod-disabled.event'
import { ReleaseService } from '../services/release.service'
import { generateSymlinkUninstallScript } from '../functions/generateSymlinkUninstallScript'
import { outputFile } from 'fs-extra'
import { join } from 'node:path'
import { compact } from 'lodash'
import { SettingsManager } from './settings.manager'

@Injectable()
export class UninstallBatManager {
  static readonly FILENAME = 'del-symlinks.bat'

  private readonly logger = new Logger(UninstallBatManager.name)

  @Inject(ReleaseService)
  private readonly releaseService: ReleaseService

  @Inject()
  private readonly settingsManager: SettingsManager

  @OnEvent(ModEnabledEvent.name)
  async onModEnabled(event: ModEnabledEvent) {
    this.logger.log(`Handling mod enabled event: ${event.modId} ${event.version}`)
    await this.rebuildUninstallBat()
  }

  @OnEvent(ModDisabledEvent.name)
  async onModDisabled(event: ModDisabledEvent) {
    this.logger.log(`Handling mod disabled event: ${event.modId} ${event.version}`)
    await this.rebuildUninstallBat()
  }

  async rebuildUninstallBat() {
    const releaseAssets = await this.releaseService.findAssetsWithSymlinks()
    const content = await generateSymlinkUninstallScript(
      compact(releaseAssets.map((it) => it.symlinkPath))
    )
    await outputFile(
      join(await this.settingsManager.getWriteDir(), UninstallBatManager.FILENAME),
      content
    )
  }
}
