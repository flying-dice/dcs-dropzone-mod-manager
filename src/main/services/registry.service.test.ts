import { Test, TestingModule } from '@nestjs/testing'
import { RegistryService } from './registry.service'
import nock from 'nock'
import { beforeAll, expect } from 'vitest'
import { ConfigModule } from '@nestjs/config'
import { MongooseModule } from '@nestjs/mongoose'
import { MongoMemoryServer } from 'mongodb-memory-server'
import { SettingsManager } from '../manager/settings.manager'
import { SettingsService } from './settings.service'
import { Config, ConfigSchema } from '../schemas/config.schema'
import registryIndex from '../__stubs__/index.json'
import modIndex from '../__stubs__/example-mod/index.json'
import { DEFAULT_REGISTRY_URL } from '../../lib/registry'

vi.mock('@aptabase/electron/main')
vi.mock('electron')

function bootstrap() {
  return Test.createTestingModule({
    imports: [
      ConfigModule.forRoot({
        load: [() => ({ registryUrl: DEFAULT_REGISTRY_URL })]
      }),
      MongooseModule.forRootAsync({
        useFactory: async () => {
          const memory = await MongoMemoryServer.create()

          return {
            uri: memory.getUri()
          }
        }
      }),
      MongooseModule.forFeature([{ name: Config.name, schema: ConfigSchema }])
    ],
    providers: [RegistryService, SettingsManager, SettingsService]
  }).compile()
}

describe('RegistryService', () => {
  describe('GIVEN default values', () => {
    let moduleRef: TestingModule

    beforeAll(async () => {
      moduleRef = await bootstrap()
    })

    it('should successfully subscribe to a mod', async () => {
      nock(DEFAULT_REGISTRY_URL).get('/index.json').reply(200, registryIndex)

      await expect(moduleRef.get(RegistryService).getRegistryIndex()).resolves.toMatchSnapshot()
    })

    it('should successfully fetch the mod index', async () => {
      nock(DEFAULT_REGISTRY_URL).get('/example-mod/index.json').reply(200, modIndex)

      await expect(
        moduleRef.get(RegistryService).getRegistryEntryIndex('example-mod')
      ).resolves.toMatchSnapshot()
    })
  })

  describe('GIVEN overridden values', () => {
    let moduleRef: TestingModule

    beforeAll(async () => {
      moduleRef = await bootstrap()

      await moduleRef.get(SettingsService).setSettingValue('registryUrl', 'https://example.com')
    })

    it('should successfully subscribe to a mod', async () => {
      nock('https://example.com').get('/index.json').reply(200, registryIndex)

      await expect(moduleRef.get(RegistryService).getRegistryIndex()).resolves.toMatchSnapshot()
    })

    it('should successfully fetch the mod index', async () => {
      nock('https://example.com').get('/example-mod/index.json').reply(200, modIndex)

      await expect(
        moduleRef.get(RegistryService).getRegistryEntryIndex('example-mod')
      ).resolves.toMatchSnapshot()
    })
  })
})
