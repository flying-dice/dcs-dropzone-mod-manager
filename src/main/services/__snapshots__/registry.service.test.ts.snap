// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegistryService > GIVEN default values > should successfully fetch the mod index 1`] = `
{
  "authors": [
    {
      "name": "Flying Dice",
    },
  ],
  "category": "example",
  "content": "IyBFeGFtcGxlIEhlbGxvIFdvcmxkIE1vZAoKRENTIHdvcmxkIEhlbGxvIFdvcmxkIE1vZAoKT24gRENTIFN0YXJ0dXAgbG9ncyBoZWxsbyB3b3JsZCB0byB0aGUgY29uc29sZQoKPiBUaGlzIGNvbnRlbnQgaXMgcHJlc2VudGVkIHRvIHRoZSB1c2VyIHdoZW4gdGhleSBvcGVuIHRoZSBtb2QgcGFnZQo=",
  "description": "A simple mod that logs hello world to the console on DCS startup",
  "homepage": "https://github.com/flying-dice/hello-world-mod",
  "id": "example-mod",
  "imageUrl": "example-mod/index.png",
  "latest": "0.1.0",
  "license": "GNU General Public License v3.0",
  "name": "Hello World Mod",
  "tags": [
    "hello",
  ],
  "versions": [
    {
      "assets": [
        {
          "source": "https://github.com/flying-dice/hello-world-mod/releases/download/0.1.0/hello-world.lua",
          "target": "{{DCS_USER_DIR}}/Scripts/Hooks/hello-world.lua",
        },
      ],
      "date": "2023-01-07T12:00:00.000Z",
      "name": "RC1",
      "releasepage": "https://github.com/flying-dice/hello-world-mod/releases/tag/0.1.0",
      "version": "0.1.0",
    },
  ],
}
`;

exports[`RegistryService > GIVEN default values > should successfully subscribe to a mod 1`] = `
[
  {
    "authors": [
      {
        "name": "Flying Dice",
      },
    ],
    "category": "example",
    "description": "A simple mod that logs hello world to the console on DCS startup",
    "id": "example-mod",
    "imageUrl": "example-mod/index.png",
    "latest": "0.1.0",
    "name": "Hello World Mod",
    "tags": [
      "hello",
    ],
  },
]
`;

exports[`RegistryService > GIVEN overridden values > should successfully fetch the mod index 1`] = `
{
  "authors": [
    {
      "name": "Flying Dice",
    },
  ],
  "category": "example",
  "content": "IyBFeGFtcGxlIEhlbGxvIFdvcmxkIE1vZAoKRENTIHdvcmxkIEhlbGxvIFdvcmxkIE1vZAoKT24gRENTIFN0YXJ0dXAgbG9ncyBoZWxsbyB3b3JsZCB0byB0aGUgY29uc29sZQoKPiBUaGlzIGNvbnRlbnQgaXMgcHJlc2VudGVkIHRvIHRoZSB1c2VyIHdoZW4gdGhleSBvcGVuIHRoZSBtb2QgcGFnZQo=",
  "description": "A simple mod that logs hello world to the console on DCS startup",
  "homepage": "https://github.com/flying-dice/hello-world-mod",
  "id": "example-mod",
  "imageUrl": "example-mod/index.png",
  "latest": "0.1.0",
  "license": "GNU General Public License v3.0",
  "name": "Hello World Mod",
  "tags": [
    "hello",
  ],
  "versions": [
    {
      "assets": [
        {
          "source": "https://github.com/flying-dice/hello-world-mod/releases/download/0.1.0/hello-world.lua",
          "target": "{{DCS_USER_DIR}}/Scripts/Hooks/hello-world.lua",
        },
      ],
      "date": "2023-01-07T12:00:00.000Z",
      "name": "RC1",
      "releasepage": "https://github.com/flying-dice/hello-world-mod/releases/tag/0.1.0",
      "version": "0.1.0",
    },
  ],
}
`;

exports[`RegistryService > GIVEN overridden values > should successfully subscribe to a mod 1`] = `
[
  {
    "authors": [
      {
        "name": "Flying Dice",
      },
    ],
    "category": "example",
    "description": "A simple mod that logs hello world to the console on DCS startup",
    "id": "example-mod",
    "imageUrl": "example-mod/index.png",
    "latest": "0.1.0",
    "name": "Hello World Mod",
    "tags": [
      "hello",
    ],
  },
]
`;
